### PR TITLE
Fix: Ignore @support block if @support or feature is not supported

### LINKED ISSUE #1869: [Feature] Compat-API should not flag issues if code is in @supports
# 🚀 Feature request

## Description

Right now the compat-api for CSS doesn't check if a feature is being featured detected previously via `@supports` and thus causing false positives.

The hint should be aware of those.

## What scenarios will this solve?

Flagging false positives when users run the compat.

### PR DESCRIPTION
**Pull request by sarvaje**
Fix #1869

## Pull request checklist

Make sure you:

- [x] Signed the [Contributor License Agreement](https://cla.js.foundation/webhintio/hint)
- [x] Followed the [commit message guidelines](https://webhint.io/docs/contributor-guide/getting-started/pull-requests/#commit-messages)

For non-trivial changes, please make sure you also:

- [ ] Added/Updated related documentation.
- [x] Added/Updated related tests.

### CODE REVIEW

File: packages/hint-compat-api/src/helpers/compat-css.ts [hunk 1]
```diff
@@ -41,14 +36,118 @@ export class CompatCSS extends CompatBase<StyleEvents, StyleParse> {
...
+            query = query.replace(`${feature.property}:${feature.value}`, featureSupported.toString());
+        }
+
+        // Ignore selector(...)
+        query = query.replace(/selector\([^)]*\)/, 'true');
+        query = query.replace(/or/gi, '||');
+        query = query.replace(/and/gi, '&&');
+        query = query.replace(/not\s*/gi, '!');
+
+        const valid = eval(query); // eslint-disable-line no-eval
```
[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/hint-compat-api/src/helpers/compat-css.ts]: I will take a look

[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/hint-compat-api/src/helpers/compat-css.ts]: @antross I make the eval "manually", let me know if you like this method, or if you prefer to use any "safe eval" package.

File: packages/hint-compat-api/src/helpers/evaluate-query.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            /*
+             * Calculate the acumulated result.
+             */
+            result = operator === '&&' ? result && partialResult : result || partialResult;
+
+            operator = null;
+        }
+    }
+
+    return result!;
```
Comment (antross) [inline packages/hint-compat-api/src/helpers/evaluate-query.ts]: If we know `result` won't be `null` here, let's include a comment explaining why. Or just `!!result` so we always ensure we return a `boolean` (if that's appropriate).

File: packages/hint-compat-api/src/helpers/compat-css.ts [hunk 3]
```diff
@@ -41,14 +37,125 @@ export class CompatCSS extends CompatBase<StyleEvents, StyleParse> {
...
+            const featureStrategy = this.chooseStrategyToSearchCSSFeature(featureNode);
+            const featureSupported = featureStrategy.testFeature(featureNode, location, true);
+
+            query = query.replace(`(${feature.property}:${feature.value})`, featureSupported.toString());
+        }
+
+        query = query.replace(/selector\s*/g, '#');
+        query = query.replace(/or/gi, '||');
+        query = query.replace(/and/gi, '&&');
+        query = query.replace(/not\s*/gi, '!');
```
[the remark this answers is not in the document]
Comment (sarvaje) [reply packages/hint-compat-api/src/helpers/compat-css.ts]: I think we don't need it anymore (but I need to make changes in the code), but it was nice having these transformation to make everything easier to follow.

**Comment by sarvaje** [comment]
@antross everything should be done :)

File: packages/hint-compat-api/src/helpers/evaluate-query.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+        finalQuery = finalQuery.replace(exec[0], evaluateQuery(condition).toString());
+
+        exec = regex.exec(finalQuery);
+    }
+
+    /*
+     * Split the query in statements.
+     */
+    const splitedQuery = finalQuery.trim().split(' ');
```
Comment (antross) [inline packages/hint-compat-api/src/helpers/evaluate-query.ts]: Naming nit.
```suggestion
    const splitQuery = finalQuery.trim().split(' ');
```

File: packages/hint-compat-api/src/helpers/evaluate-query.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+         */
+        if (result === null) {
+            result = partialResult;
+
+            continue;
+        }
+
+        if (operator) {
+            /*
+             * Calculate the acumulated result.
```
Comment (antross) [inline packages/hint-compat-api/src/helpers/evaluate-query.ts]:
```suggestion
             * Calculate the accumulated result.
```

File: packages/hint-compat-api/src/types.ts [hunk 6]
```diff
@@ -9,12 +9,24 @@ export type BrowserSupportCollection = {
     [key: string]: number[];
 };
 
+export type TestFeatureOptions = {
+    /**
+     * Indicate if the report of a feature has to be skiped
```
Comment (antross) [inline packages/hint-compat-api/src/types.ts]:
```suggestion
     * Indicate if the report of a feature has to be skipped
```

File: packages/hint-compat-api/src/types.ts [hunk 7]
```diff
@@ -9,12 +9,24 @@ export type BrowserSupportCollection = {
...
+     * This is necessary because in some cases, we just need to
+     * check if a feature is valid or not, without reporting it.
```
Comment (antross) [inline packages/hint-compat-api/src/types.ts]:
```suggestion
     * check if a feature is valid or not, without reporting it.
     * E.g. when deciding whether to ignore a `@supports` block.
```

**Comment by sarvaje** [comment]
@antross done!

**Pull request merged**